src/System.Slices/System/Buffers/Memory.cs

```diff
@@ -55,7 +55,7 @@ public Memory<T> Slice(int index, int length)
 
         public Span<T> Span => _owner.GetSpanInternal(_id).Slice(_index, _length);
 
-        public DisposableReservation Reserve() => new DisposableReservation(_owner, _id);
+        public DisposableReservation<T> Reserve() => new DisposableReservation<T>(_owner, _id);
```

Comment:
The idea was that the pipeline would store tokens for all buffer segments,a nd then dispose them all when it's done with the buffer. But I just realized in such scenario, all tokens will be Disposable<byte>.

Let's just make it generic then and see how it feels to use it.